File: 20-actors/shionome/methods/grounding.py
"""grounding.py — 潮目 (shionome) stock-layer ENTITY-GROUNDING bridge. ADR-2606072200 (R1).

The stock pyramid (weave.stock_pyramid) sizes each money-and-markets LAYER as an aggregate
(equities ≈ $115tn, …). This bridge answers the next question HONESTLY: *who is inside each
layer?* — it decomposes a pyramid layer into the NAMED real entities that sibling actors already
mirror, and reports the COVERAGE gap (how much of the layer is actually grounded vs still
illustrative).

  - the EQUITIES layer ← kabuto 兜 listed-company ledger (org.corp.* — name/ticker/market-cap)
  - a cross-cutting SYSTEMIC-INSTITUTIONS overlay ← hokorobi 綻び (G-SIB banks / insurers /
    pensions / CCPs span equities+debt+pensions, so they are an overlay, NOT one pyramid layer)

DISCIPLINE (unchanged from the actor's gates):
  - G2 トレードはしない — a market-cap / institution count is a SIZE, descriptive, never a
    rating / signal / target / solvency verdict. Every report carries no_trade_notice = true.
  - G1 — the grounded entities are PUBLIC companies / systemic institutions already mirrored by
    kabuto / hokorobi (public/power-only, person-excluded); shionome adds no new entity, only a
    read-side view that names the constituents of a public asset class.
  - HONESTY — value_coverage is a stated LOWER BOUND (only sized companies contribute); the
    count denominator is an explicit :representative universe figure, not a live count;
    ungrounded_layers names exactly what is NOT yet backed by entities.
  - FAIL-OPEN — a missing/unreadable sibling ledger yields [] (the layer is reported ungrounded),
    never a crash. shionome's core (weave/concentration) does NOT import this bridge.

Stdlib only. Pure functions take already-loaded record lists; only load_ledger touches disk.
"""

from __future__ import annotations

import pathlib
from collections import Counter

from weave import _kw

# A :representative universe denominator (a documented public order-of-magnitude reference, NOT a
# live count) used ONLY to express count coverage as an honest fraction. ~50k–58k companies are
# exchange-listed worldwide (WFE statistics order of magnitude).
LISTED_UNIVERSE = 55000


def load_ledger(path) -> list:
    """Load a sibling actor's EDN ledger (a top-level vector of records). FAIL-OPEN: a missing or
    unreadable ledger returns [] — the bridge degrades to 'ungrounded', it never crashes shionome."""
    p = pathlib.Path(path)
    if not p.exists():
        return []
    try:
        from _edn import load_edn
        data = load_edn(p)
    except Exception:  # noqa: BLE001 — fail-open by design
        return []
    return data if isinstance(data, list) else []


def kabuto_equity_constituents(records: list) -> list[dict]:
    """Named listed-company constituents of the global EQUITIES layer, from a kabuto ledger. A
    company record carries :company/name (address / contact / supply-edge / process records do
    not, so they are skipped). :company/market-cap-busd is a FACTUAL public-record size, never a
    rating (G2); it may be absent (the company is then named-but-unsized)."""
    out = []
    for r in records:
        if not isinstance(r, dict) or not r.get(":company/name"):
            continue
        mc = r.get(":company/market-cap-busd")
        out.append({
            "id": r.get(":company/id"),
            "name": r.get(":company/name"),
            "ticker": r.get(":company/ticker"),
            "country": r.get(":company/country"),
            "sector": _kw(r.get(":company/sector")) if r.get(":company/sector") else None,
            "market_cap_busd": float(mc) if mc is not None else None,
            "sourcing": _kw(r.get(":company/sourcing")) if r.get(":company/sourcing") else None,
        })
    return out


def hokorobi_institutions(records: list) -> list[dict]:
    """Systemic financial INSTITUTIONS from a hokorobi ledger (:organism/kind :institution). A
    cross-cutting overlay (G-SIB banks / insurers / pensions / CCPs span equities+debt+pensions —
    they do not belong to a single pyramid layer). Resilience overlay, never a solvency verdict (G2)."""
    out = []
    for r in records:
        if not isinstance(r, dict) or _kw(r.get(":organism/kind")) != "institution":
            continue
        out.append({
            "id": r.get(":organism/id"),
            "label": r.get(":organism/label"),
            "sector": _kw(r.get(":inst/sector")) if r.get(":inst/sector") else None,
            "sii": _kw(r.get(":inst/sii")) if r.get(":inst/sii") else None,
            "jurisdiction": r.get(":inst/jurisdiction"),
            "sourcing": _kw(r.get(":organism/sourcing")) if r.get(":organism/sourcing") else None,
        })
    return out


def ground_equities(layer_usd_tn: float, constituents: list[dict], universe: int = LISTED_UNIVERSE) -> dict:
    """Ground the EQUITIES stock layer in named listed companies, reporting HONESTLY:
      - grounded_entities       — how many named companies back the layer,
      - grounded_market_cap_usd_tn — Σ market-cap of the sized subset (→ value coverage of $layer),
      - value_coverage_of_layer — a LOWER BOUND (only companies that report a market-cap count),
      - count_coverage_of_universe — fraction of a :representative listed-universe denominator.
    A size is factual; nothing here is a rating/signal/target (G2)."""
    n = len(constituents)
    sized = [c for c in constituents if c["market_cap_busd"] is not None]
    mcap_tn = sum(c["market_cap_busd"] for c in sized) / 1000.0  # USD billions → USD trillions
    top = sorted(sized, key=lambda c: -c["market_cap_busd"])[:10]
    return {
        "layer": "equities",
        "layer_usd_tn": round(float(layer_usd_tn), 4),
        "grounded_entities": n,
        "entities_with_size": len(sized),
        "grounded_market_cap_usd_tn": round(mcap_tn, 4),
        "value_coverage_of_layer": round(mcap_tn / layer_usd_tn, 4) if layer_usd_tn else 0.0,
        "value_coverage_is_lower_bound": len(sized) < n,
        "count_coverage_of_universe": round(n / universe, 5) if universe else 0.0,
        "universe_denominator": universe,
        "universe_sourcing": "representative",   # WFE order-of-magnitude, NOT a live count
        "top_constituents": [
            {"name": c["name"], "ticker": c["ticker"], "market_cap_busd": c["market_cap_busd"]}
            for c in top
        ],
        "no_trade_notice": True,
    }


def systemic_overlay(institutions: list[dict]) -> dict:
    """The cross-cutting systemic-institution overlay (from hokorobi). Counts by sector + sourcing
    honesty; it sizes nothing against a single layer. Resilience map, never a solvency verdict (G2)."""
    by_sector = dict(sorted(Counter(i["sector"] or "(unknown)" for i in institutions).items()))
    auth = sum(1 for i in institutions if i["sourcing"] == "authoritative")
    return {
        "institutions": len(institutions),
        "by_sector": by_sector,
        "authoritative": auth,
        "representative": len(institutions) - auth,
        "note": "cross-cutting systemic overlay — G-SIB banks / insurers / pensions / CCPs span "
                "equities+debt+pensions, not one pyramid layer; resilience map, never a solvency verdict",
        "no_trade_notice": True,
    }


def ground(pyramid: dict, kabuto_records: list, hokorobi_records: list) -> dict:
    """The full stock-layer grounding report: decompose pyramid layers into named real entities
    WHERE a sibling ledger exists, and stay HONEST about the rest (ungrounded_layers). Every figure
    is an aggregate / size / coverage fraction — never a per-entity rating/signal/target (G2/G4)."""
    layers = {l["asset_class"]: l for l in pyramid.get("layers", [])}
    eq_layer = float(layers.get("equities", {}).get("usd_tn", 0.0))
    equities = ground_equities(eq_layer, kabuto_equity_constituents(kabuto_records))
    overlay = systemic_overlay(hokorobi_institutions(hokorobi_records))
    grounded_classes = {"equities"} if equities["grounded_entities"] else set()
    ungrounded = [l["asset_class"] for l in pyramid.get("layers", [])
                  if l["asset_class"] not in grounded_classes]
    return {
        "equities": equities,
        "systemic_institutions_overlay": overlay,
        "ungrounded_layers": ungrounded,
        "summary": {
            "pyramid_layers": len(pyramid.get("layers", [])),
            "layers_with_entity_grounding": len(grounded_classes),
            "total_named_entities": equities["grounded_entities"] + overlay["institutions"],
            "no_trade_notice": True,
        },
    }


if __name__ == "__main__":
    from _edn import load_edn
    from weave import stock_pyramid, weave

    here = pathlib.Path(__file__).resolve()
    actor = here.parents[1]
    root = here.parents[3]
    g = weave(load_edn(actor / "data" / "seed-capital-flow-graph.kotoba.edn"))
    pyr = stock_pyramid(g)
    kab = load_ledger(root / "20-actors" / "kabuto" / "data" / "seed-public-companies.kotoba.edn")
    hok = load_ledger(root / "20-actors" / "hokorobi" / "data" / "seed-finrisk-graph.kotoba.edn")
    rep = ground(pyr, kab, hok)

    eq = rep["equities"]
    print("# 潮目 (shionome) — stock-layer ENTITY GROUNDING (どこまで実エンティティで裏付けたか)\n")
    print(f"## equities layer — ${eq['layer_usd_tn']:.0f}tn, grounded by {eq['grounded_entities']} named companies")
    print(f"- value coverage: ${eq['grounded_market_cap_usd_tn']:.1f}tn of ${eq['layer_usd_tn']:.0f}tn "
          f"= {eq['value_coverage_of_layer']*100:.1f}%"
          + ("  (LOWER BOUND — only sized companies count)" if eq["value_coverage_is_lower_bound"] else ""))
    print(f"- count coverage: {eq['grounded_entities']} / ~{eq['universe_denominator']} listed "
          f"(:{eq['universe_sourcing']}) = {eq['count_coverage_of_universe']*100:.2f}%")
    print("- top constituents:")
    for c in eq["top_constituents"]:
        mc = c["market_cap_busd"]
        print(f"    {c['name']} ({c['ticker']}): ${mc:.0f}bn" if mc is not None else f"    {c['name']}")
    ov = rep["systemic_institutions_overlay"]
    print(f"\n## systemic-institutions overlay — {ov['institutions']} institutions "
          f"({ov['authoritative']} authoritative): {ov['by_sector']}")
    print(f"\n## ungrounded layers (NOT yet backed by named entities): {rep['ungrounded_layers']}")
    s = rep["summary"]
    print(f"\n## summary: {s['layers_with_entity_grounding']}/{s['pyramid_layers']} pyramid layers grounded; "
          f"{s['total_named_entities']} named entities total — descriptive, NOT advice (トレードはしない)")

File: 20-actors/shionome/methods/grounding.clj
#!/usr/bin/env bb
;; Clojure port of methods/grounding.py — 潮目 (shionome) stock-layer entity-grounding bridge.
(ns shionome.methods.grounding
  "grounding.clj — 潮目 (shionome) stock-layer ENTITY-GROUNDING bridge. ADR-2606072200 (R1).

  The stock pyramid (weave/stock-pyramid) sizes each money-and-markets LAYER as an aggregate.
  This bridge answers the next question HONESTLY: *who is inside each layer?* — it decomposes a
  pyramid layer into the NAMED real entities that sibling actors already mirror, and reports the
  COVERAGE gap (how much of the layer is actually grounded vs still illustrative).

    - the EQUITIES layer ← kabuto 兜 listed-company ledger (org.corp.* — name/ticker/market-cap)
    - a cross-cutting SYSTEMIC-INSTITUTIONS overlay ← hokorobi 綻び (G-SIB banks / insurers /
      pensions / CCPs span equities+debt+pensions, so they are an overlay, NOT one pyramid layer)

  DISCIPLINE (unchanged from the actor's gates):
    - G2 トレードはしない — a market-cap / institution count is a SIZE, descriptive, never a
      rating / signal / target / solvency verdict. Every report carries no_trade_notice = true.
    - G1 — the grounded entities are PUBLIC companies / systemic institutions already mirrored by
      kabuto / hokorobi; shionome adds no new entity, only a read-side view.
    - HONESTY — value_coverage is a stated LOWER BOUND; the count denominator is a :representative
      universe figure, not a live count; ungrounded_layers names exactly what is NOT yet backed.
    - FAIL-OPEN — a missing/unreadable sibling ledger yields [] (the layer is reported ungrounded).

  Stdlib only. Pure functions take already-loaded record lists; only load-ledger touches disk."
  (:require [clojure.string :as str]
            [shionome.methods.edn :as sedn]
            [clojure.java.io :as io]))

;; ── kw* inlined from weave.cljc (private there — babashka cannot reach cross-ns private vars) ──
;; Mirror of weave.cljc `kw*` which mirrors weave.py `_kw`:
;;   str(v or "").lstrip(":") → split("/")[-1].lower()
(defn- kw*
  "Normalize an edn keyword/string to a bare lowercase token (':flow/kind' → 'kind').
  Inlined mirror of weave.cljc `kw*` (private there). Do NOT modify independently."
  [v]
  (let [s (-> (str (or v "")) (str/replace #"^:+" ""))]
    (-> (last (str/split s #"/" -1)) str/lower-case)))

;; A :representative universe denominator used ONLY to express count coverage as an honest
;; fraction. ~50k–58k companies are exchange-listed worldwide (WFE statistics order of magnitude).
(def LISTED_UNIVERSE 55000)

;; ── per-layer grounding ROADMAP ──────────────────────────────────────────────────────────────────
;; For every money-pyramid asset-class: which sibling actor's entity ledger (if any) can decompose
;; it, and — where it CANNOT yet — the explicit reason.
(def LAYER_GROUNDING
  {"equities"    {"source_actor" "kabuto"
                  "status"       "grounded"
                  "reason"       "listed-company ledger (org.corp.*) decomposes the layer by named issuer; kanjō adds disclosure depth"}
   "debt"        {"source_actor" nil
                  "status"       "ungroundable-at-r0"
                  "reason"       "bond-market SIZE is a SIFMA/BIS aggregate; no per-issuer debt-outstanding ledger (kanjō discloses corporate BS, NOT tradable debt-securities outstanding — conflating them would mis-size the layer). Candidate issuers: ooyake (sovereign) + kabuto (corporate)"}
   "broad-money" {"source_actor" nil
                  "status"       "ungroundable-at-r0"
                  "reason"       "M2 is a central-bank / banking-system aggregate (BIS / World Bank); not entity-decomposable"}
   "cash"        {"source_actor" nil
                  "status"       "ungroundable-at-r0"
                  "reason"       "physical currency outstanding is a central-bank aggregate; no holder ledger (G1 bars holder-tracking)"}
   "real-estate" {"source_actor" nil
                  "status"       "ungroundable-at-r0"
                  "reason"       "no real-estate-entity / parcel ledger in the repo; total value is a Savills aggregate"}
   "gold"        {"source_actor" nil
                  "status"       "ungroundable-at-r0"
                  "reason"       "above-ground stock value is a World Gold Council aggregate; no holder ledger (G1 bars holder-tracking)"}
   "crypto"      {"source_actor" nil
                  "status"       "ungroundable-at-r0"
                  "reason"       "total market-cap aggregate; no issuer ledger in the repo"}
   "derivatives" {"source_actor" nil
                  "status"       "ungroundable-at-r0"
                  "reason"       "OTC gross-notional is a BIS aggregate; not entity-decomposable from any repo ledger"}})

;; ── helpers ──────────────────────────────────────────────────────────────────────────────────────

(defn- round-n
  "Round x to n decimal places (matches Python round(x, n))."
  [x n]
  (/ (Math/round (* (double x) (Math/pow 10 n))) (Math/pow 10 n)))

(defn load-ledger
  "Load a sibling actor's EDN ledger (a top-level vector of records). FAIL-OPEN: a missing or
  unreadable ledger returns [] — the bridge degrades to 'ungrounded', it never crashes shionome."
  [path]
  (let [f (io/file (str path))]
    (if-not (.exists f)
      []
      (try
        (let [data (sedn/load-edn f)]
          (if (sequential? data) (vec data) []))
        (catch Exception _ [])))))

;; ── constituent extraction ────────────────────────────────────────────────────────────────────────

(defn kabuto-equity-constituents
  "Named listed-company constituents of the global EQUITIES layer, from a kabuto ledger. A
  company record carries :company/name (address / contact / supply-edge / process records do
  not, so they are skipped). :company/market-cap-busd is a FACTUAL public-record size, never a
  rating (G2); it may be absent (the company is then named-but-unsized)."
  [records]
  (reduce
   (fn [out r]
     (if (and (map? r) (get r ":company/name"))
       (let [mc (get r ":company/market-cap-busd")]
         (conj out
               {"id"             (get r ":company/id")
                "name"           (get r ":company/name")
                "ticker"         (get r ":company/ticker")
                "country"        (get r ":company/country")
                "sector"         (when (get r ":company/sector")
                                   (kw* (get r ":company/sector")))
                "market_cap_busd" (when (some? mc) (double mc))
                "sourcing"       (when (get r ":company/sourcing")
                                   (kw* (get r ":company/sourcing")))}))
       out))
   []
   records))

(defn hokorobi-institutions
  "Systemic financial INSTITUTIONS from a hokorobi ledger (:organism/kind :institution). A
  cross-cutting overlay (G-SIB banks / insurers / pensions / CCPs span equities+debt+pensions).
  Resilience overlay, never a solvency verdict (G2)."
  [records]
  (reduce
   (fn [out r]
     (if (and (map? r) (= (kw* (get r ":organism/kind")) "institution"))
       (conj out
             {"id"           (get r ":organism/id")
              "label"        (get r ":organism/label")
              "sector"       (when (get r ":inst/sector")
                               (kw* (get r ":inst/sector")))
              "sii"          (when (get r ":inst/sii")
                               (kw* (get r ":inst/sii")))
              "jurisdiction" (get r ":inst/jurisdiction")
              "sourcing"     (when (get r ":organism/sourcing")
                               (kw* (get r ":organism/sourcing")))})
       out))
   []
   records))

(defn kanjo-disclosed-companies
  "The set of org.corp.* company ids that have ≥1 DISCLOSED filing in a kanjō ledger
  (:fin.filing/company). Used to measure DISCLOSURE DEPTH of the equities layer."
  [records]
  (reduce
   (fn [out r]
     (if (and (map? r) (get r ":fin.filing/company"))
       (conj out (get r ":fin.filing/company"))
       out))
   #{}
   records))

;; ── grounding math ────────────────────────────────────────────────────────────────────────────────

(defn ground-equities
  "Ground the EQUITIES stock layer in named listed companies, reporting HONESTLY."
  ([layer-usd-tn constituents] (ground-equities layer-usd-tn constituents LISTED_UNIVERSE nil))
  ([layer-usd-tn constituents universe] (ground-equities layer-usd-tn constituents universe nil))
  ([layer-usd-tn constituents universe disclosed-ids]
   (let [disclosed-ids  (or disclosed-ids #{})
         n              (count constituents)
         sized          (filter #(some? (get % "market_cap_busd")) constituents)
         sized-v        (vec sized)
         mcap-tn        (/ (reduce + 0.0 (map #(double (get % "market_cap_busd")) sized-v))
                           1000.0)    ; USD billions → USD trillions
         top            (vec (take 10 (sort-by #(- (double (get % "market_cap_busd"))) sized-v)))
         disclosed-cons (filter #(contains? disclosed-ids (get % "id")) constituents)]
     {"layer"                      "equities"
      "layer_usd_tn"               (round-n (double layer-usd-tn) 4)
      "grounded_entities"          n
      "entities_with_size"         (count sized-v)
      "grounded_market_cap_usd_tn" (round-n mcap-tn 4)
      "value_coverage_of_layer"    (if (> (double layer-usd-tn) 0.0)
                                     (round-n (/ mcap-tn (double layer-usd-tn)) 4)
                                     0.0)
      "value_coverage_is_lower_bound" (< (count sized-v) n)
      "count_coverage_of_universe" (if (> (int universe) 0)
                                     (round-n (/ (double n) (double universe)) 5)
                                     0.0)
      "universe_denominator"       universe
      "universe_sourcing"          "representative"   ; WFE order-of-magnitude, NOT a live count
      "with_disclosed_financials"  (count disclosed-cons)
      "disclosed_sample"           (vec (take 10 (map #(get % "name") disclosed-cons)))
      "top_constituents"           (mapv (fn [c] {"name"           (get c "name")
                                                   "ticker"         (get c "ticker")
                                                   "market_cap_busd" (get c "market_cap_busd")})
                                         top)
      "no_trade_notice"            true})))

(defn grounding-roadmap
  "Per-layer grounding STATUS for every money-pyramid layer — the honest map of what is
  entity-grounded vs what cannot be (and why), from the LAYER_GROUNDING registry."
  [pyramid]
  (mapv (fn [l]
          (let [ac   (get l "asset_class")
                spec (get LAYER_GROUNDING ac
                          {"source_actor" nil
                           "status"       "unmapped"
                           "reason"       "no grounding spec for this asset class"})]
            {"asset_class"   ac
             "layer_usd_tn"  (get l "usd_tn")
             "source_actor"  (get spec "source_actor")
             "status"        (get spec "status")
             "reason"        (get spec "reason")}))
        (get pyramid "layers" [])))

(defn systemic-overlay
  "The cross-cutting systemic-institution overlay (from hokorobi). Counts by sector + sourcing
  honesty; it sizes nothing against a single layer. Resilience map, never a solvency verdict (G2)."
  [institutions]
  (let [by-sector (reduce (fn [m i]
                            (let [s (or (get i "sector") "(unknown)")]
                              (update m s (fnil inc 0))))
                          (sorted-map)
                          institutions)
        auth      (count (filter #(= (get % "sourcing") "authoritative") institutions))]
    {"institutions"  (count institutions)
     "by_sector"     by-sector
     "authoritative" auth
     "representative" (- (count institutions) auth)
     "note"          (str "cross-cutting systemic overlay — G-SIB banks / insurers / pensions / CCPs span "
                          "equities+debt+pensions, not one pyramid layer; resilience map, never a solvency verdict")
     "no_trade_notice" true}))

(defn ground
  "The full stock-layer grounding report: decompose pyramid layers into named real entities
  WHERE a sibling ledger exists, add disclosure DEPTH (kanjō) to the equities layer, and stay
  HONEST about the rest via the per-layer roadmap."
  ([pyramid kabuto-records hokorobi-records]
   (ground pyramid kabuto-records hokorobi-records nil))
  ([pyramid kabuto-records hokorobi-records kanjo-records]
   (let [layers         (into {} (map (juxt #(get % "asset_class") identity)
                                      (get pyramid "layers" [])))
         eq-layer       (double (get (get layers "equities" {}) "usd_tn" 0.0))
         disclosed      (kanjo-disclosed-companies (or kanjo-records []))
         constituents   (kabuto-equity-constituents kabuto-records)
         equities       (ground-equities eq-layer constituents LISTED_UNIVERSE disclosed)
         overlay        (systemic-overlay (hokorobi-institutions hokorobi-records))
         grounded-cls   (if (> (get equities "grounded_entities") 0) #{"equities"} #{})
         ungrounded     (vec (keep (fn [l]
                                     (let [ac (get l "asset_class")]
                                       (when-not (contains? grounded-cls ac) ac)))
                                   (get pyramid "layers" [])))]
     {"equities"                       equities
      "systemic_institutions_overlay"  overlay
      "roadmap"                        (grounding-roadmap pyramid)
      "ungrounded_layers"              ungrounded
      "summary"                        {"pyramid_layers"               (count (get pyramid "layers" []))
                                        "layers_with_entity_grounding" (count grounded-cls)
                                        "total_named_entities"         (+ (get equities "grounded_entities")
                                                                          (get overlay "institutions"))
                                        "no_trade_notice"              true}})))
